examples/fluids/qfunctions/utils.h

```diff
@@ -18,17 +18,128 @@
 CEED_QFUNCTION_HELPER CeedScalar Max(CeedScalar a, CeedScalar b) { return a < b ? b : a; }
 CEED_QFUNCTION_HELPER CeedScalar Min(CeedScalar a, CeedScalar b) { return a < b ? a : b; }
 
+CEED_QFUNCTION_HELPER void SwapScalar(CeedScalar *a, CeedScalar *b) {
+  CeedScalar temp = *a;
+  *a              = *b;
+  *b              = temp;
+}
+
 CEED_QFUNCTION_HELPER CeedScalar Square(CeedScalar x) { return x * x; }
 CEED_QFUNCTION_HELPER CeedScalar Cube(CeedScalar x) { return x * x * x; }
 
 // @brief Scale vector of length N by scalar alpha
 CEED_QFUNCTION_HELPER void ScaleN(CeedScalar *u, const CeedScalar alpha, const CeedInt N) {
-  CeedPragmaSIMD for (CeedInt i = 0; i < N; i++) { u[i] *= alpha; }
+  CeedPragmaSIMD for (CeedInt i = 0; i < N; i++) u[i] *= alpha;
+}
+
+// @brief Set vector of length N to a value alpha
+CEED_QFUNCTION_HELPER void SetValueN(CeedScalar *u, const CeedScalar alpha, const CeedInt N) {
+  CeedPragmaSIMD for (CeedInt i = 0; i < N; i++) u[i] = alpha;
+}
+
+// @brief Copy N elements from x to y
+CEED_QFUNCTION_HELPER void CopyN(const CeedScalar *x, CeedScalar *y, const CeedInt N) { CeedPragmaSIMD for (CeedInt i = 0; i < N; i++) y[i] = x[i]; }
+
+// @brief Copy 3x3 matrix from A to B
+CEED_QFUNCTION_HELPER void CopyMat3(const CeedScalar A[3][3], CeedScalar B[3][3]) { CopyN((const CeedScalar *)A, (CeedScalar *)B, 9); }
+
+// @brief Dot product of vectors with N elements
+CEED_QFUNCTION_HELPER CeedScalar DotN(const CeedScalar *u, const CeedScalar *v, const CeedInt N) {
+  CeedScalar output = 0;
+  CeedPragmaSIMD for (CeedInt i = 0; i < N; i++) output += u[i] * v[i];
+  return output;
 }
 
 // @brief Dot product of 3 element vectors
 CEED_QFUNCTION_HELPER CeedScalar Dot3(const CeedScalar u[3], const CeedScalar v[3]) { return u[0] * v[0] + u[1] * v[1] + u[2] * v[2]; }
 
+// @brief Cross product of vectors with 3 elements
+CEED_QFUNCTION_HELPER void Cross3(const CeedScalar u[3], const CeedScalar v[3], CeedScalar w[3]) {
+  w[0] = (u[1] * v[2]) - (u[2] * v[1]);
+  w[1] = (u[2] * v[0]) - (u[0] * v[2]);
+  w[2] = (u[0] * v[1]) - (u[1] * v[0]);
+}
+
+// @brief Curl of vector given its gradient
+CEED_QFUNCTION_HELPER void Curl3(const CeedScalar gradient[3][3], CeedScalar v[3]) {
+  v[0] = gradient[2][1] - gradient[1][2];
+  v[1] = gradient[0][2] - gradient[2][0];
+  v[2] = gradient[1][0] - gradient[0][1];
+}
+
+// @brief Matrix vector product, b = Ax + b. A is NxM, x is M, b is N
+CEED_QFUNCTION_HELPER void MatVecNM(const CeedScalar *A, const CeedScalar *x, const CeedInt N, const CeedInt M, const CeedTransposeMode transpose_A,
```

Comment:
I assume using `CeedTransposeMode` here is ok? I originally made my own `enum`, but then realized that it was functionally identical to this one.

Reply:
Looks like there might be a problem for CUDA backends (and probably HIP):
```
/home/jawr8143/software/libCEED/backends/cuda/ceed-cuda-compile.cpp:93 in CeedCompileCuda(): NVRTC_ERROR_COMPILATION
default_program(422): error: identifier "CeedTransposeMode" is undefined

default_program(425): error: identifier "CEED_NOTRANSPOSE" is undefined

default_program(428): error: identifier "CEED_TRANSPOSE" is undefined
[...]
```

If I follow the code correctly, `CeedTransposeMode` is only defined in `<ceed.h>`, but that header gets stripped away when loading in the JIT source. 

Should we just add the `CeedTransposeMode` definition to `ceed/jit-source/cuda/cuda-jit.h` and it's HIP equivalent?

Reply:
If we want to move it (should be fine to do so), then it should go into `<ceed/types.h>`

Reply:
Yep, the following change fixes it:
```diff
diff --git i/include/ceed/ceed.h w/include/ceed/ceed.h
index 233200cf..a838f477 100644
--- i/include/ceed/ceed.h
+++ w/include/ceed/ceed.h
@@ -258,14 +258,7 @@ CEED_EXTERN const CeedElemRestriction CEED_ELEMRESTRICTION_NONE;
 /// @ingroup CeedQFunction
 CEED_EXTERN const CeedQFunction CEED_QFUNCTION_NONE;

-/// Denotes whether a linear transformation or its transpose should be applied
-/// @ingroup CeedBasis
-typedef enum {
-  /// Apply the linear transformation
-  CEED_NOTRANSPOSE,
-  /// Apply the transpose
-  CEED_TRANSPOSE
-} CeedTransposeMode;
+/// Maps enum CeedTransposeMode to strings
 CEED_EXTERN const char *const CeedTransposeModes[];

 /// Argument for CeedElemRestrictionCreateStrided that L-vector is in the Ceed backend's preferred layout.
diff --git i/include/ceed/types.h w/include/ceed/types.h
index b7d32010..650e616a 100644
--- i/include/ceed/types.h
+++ w/include/ceed/types.h
@@ -137,4 +137,13 @@ typedef enum {
   CEED_ERROR_UNSUPPORTED = -3,
 } CeedErrorType;

+/// Denotes whether a linear transformation or its transpose should be applied
+/// @ingroup CeedBasis
+typedef enum {
+  /// Apply the linear transformation
+  CEED_NOTRANSPOSE,
+  /// Apply the transpose
+  CEED_TRANSPOSE
+} CeedTransposeMode;
+
 #endif
```

Have to keep the ` CEED_EXTERN const char *const CeedTransposeModes[];` in `ceed.h` unless we want to move `CEED_EXTERN` inside of `ceed/types.h` (or some other option).